### ISSUE TITLE
npm link causes EISDIR: illegal operation on a directory, read

### ISSUE THREAD

**Issue by kennu** [serverless/serverless#896]
I'm trying to deploy using Serverless 0.5.1. My component structure is like this:

    users/node_modules
    users/package.json
    users/s-templates.json
    users/users-login
    users/users-register

Deploying the functions works fine when node_modules is installed with npm install. But when I use npm link <module> to use linked modules during development, I get these errors from Serverless:

    Serverless: Deploying functions in "development" to the following regions: eu-west-1  
    Serverless: ------------------------  
    Serverless: Failed to deploy the following functions in "development" to the following regions:  
    Serverless: eu-west-1 ------------------------  
    Serverless:   users-register: EISDIR: illegal operation on a directory, read  
    Serverless:   users-login: EISDIR: illegal operation on a directory, read  

It seems like something is incompatible with symlinked modules.

**Comment by HyperBrain** [comment]
Related to #866. The CodePackageLambda action already follows symlinks during packaging.
Does `npm link` create hard or softlinks?

**Comment by HyperBrain** [comment]
#866

**Comment by kennu** [comment]
They are softlinks, in my case e.g.

    lrwxr-xr-x  1 kennu  staff  69 Mar 26 15:04 users/node_modules/messai-models -> /Users/kennu/.nvm/versions/node/v4.3.1/lib/node_modules/messai-models

**Comment by eahefnawy** [comment]
@HyperBrain I don't think your PR on the related issue was released yet. Should your PR fix this particular issue too?

**Comment by edclement** [comment]
yeah, I've run into this bug on windows as well.

**Comment by HyperBrain** [comment]
@eahefnawy My PR adds the following of symlinks to the code packaging/copy directories (in Runtime.js). So I assume that everything residing in symlinked directories should be copied.

**Comment by eahefnawy** [comment]
hmmm I was able to reproduce this issue. Looking at debug mode shows that it's coming from the wrench operation in the `Runtime` file, even though it's configured to follow symbolic links thanks to @HyperBrain PR. So it's probably something else.

@HyperBrain are symbolic links still working for you? maybe it's our npm version?

**Comment by HyperBrain** [comment]
@eahefnawy I have to check. Did we update the wrench version or is that fixed? Could also be a possible reason.

**Comment by HyperBrain** [comment]
@eahefnawy I did not check yet, but I checked the npm and GitHub for wrench. It seems the project is pretty dead with some issues open. Maybe it would make more sense to exchange it with a suitable and maintained module.

**Comment by eahefnawy** [comment]
@HyperBrain I guess you're right. Wrench is getting even weirder! I tried removing the `inflateSymlinks: true` option from `Runtime`, and it's now it's working!! It's like it's doing the reverse of what it should be doing. I confirmed this by checking the generated `tmp` directory and it's including the module I linked.

I'm gonna push this fix, even though it doesn't make sense, but I'd love if @kennu and @HyperBrain to test it out to make sure it's working for both of you.

**Comment by ryanmcgrath** [comment]
Hi-ho, author of Wrench here. Someone reached out about taking over the project but wanted to drop a note here and apologize for the state it's in - I haven't maintained it because I built it for the early days of Node, and other projects surpassed it. I've kept it around because so many projects included it, and well... removing a package entirely and screwing over the ecosystem just seems like a screwball move. I put out a call on Twitter a few times for new maintainers but I probably could have done a better job of finding one.

tl;dr The lack of free time to maintain projects on GitHub is a villain we all sadly never escape the constraints of. Absolutely open to solutions.

**Comment by ac360** [comment]
Hi @ryanmcgrath - Thank you for your great work with Wrench.  I've used it in a lot of projects over the years, as well as the Serverless Framework.  I think we're going to switch over to `fs-extra` because we currently have a lot on our plate.  But wanted to let you know that your work is appreciated and it has had a positive impact on our lives :)

**Comment by ryanmcgrath** [comment]
@ac360 No worries! `fs-extra` is what I recommend people switch to anyway. I'll try to find time this weekend to mark wrench as deprecated, at the very least (unless some magic person appears to take it over, but in all honesty I think one standard community solution is better anyway - perhaps that's just the Pythonista in me speaking).

**Comment by ryanmcgrath** [comment]
Apologies for hijacking a `serverless` thread, but just continuing off my last comment; went ahead and deprecated `wrench` now. Deprecation notice pushes people towards `fs-extra`. If anyone wants to take it over they're welcome to, but this seems good to do for now. Thanks!

**Comment by eahefnawy** [comment]
@ryanmcgrath Thanks for your great work. Always appreciated :)

**Comment by eahefnawy** [comment]
As final remarks on this issue, we've switched to `fs-extra` and I no longer see this error. @kennu @HyperBrain would love if you guys confirm so we can close this :blush:

**Comment by kennu** [comment]
Unfortunately I'm seeing errors like this, when I deploy a function that uses symlinked Node modules and try to run it:

> Unable to import module 'users-create/_serverless_handler': Error at Function.Module._resolveFilename (module.js:338:15) at Function.Module._load (module.js:280:25) at Module.require (module.js:364:17) at require (module.js:380:17) at Object.<anonymous> (/var/task/users-create/handler.js:10:13) at Module._compile (module.js:456:26) at Object.Module._extensions..js (module.js:474:10) at Module.load (module.js:356:32) at Function.Module._load (module.js:312:12) at Module.require (module.js:364:17)

When I remove the symlinked module and install it "normally", everything works fine.

Is there an easy way to check the contents of the deployed Lambda function?

**Comment by HyperBrain** [comment]
if you run `function deploy` with the `-t` parameter the packaged code is found in `meta/_tmp`

**Comment by kennu** [comment]
Okay thanks, I tested with -t. The problem seems to be that my linked npm package has a relative symlink, and when it's moved under _meta/_tmp, it doesn't point to the right target any more:

    ll _meta/_tmp/root-get\@1459757492934/node_modules/
    lrwxr-xr-x  1 kennu  staff   65 Apr  4 11:11 accomodels -> ../../../../.nvm/versions/node/v4.4.2/lib/node_modules/accomodels

    ll _meta/_tmp/root-get\@1459757492934/node_modules/accomodels/
    ls: _meta/_tmp/root-get@1459757492934/node_modules/accomodels/: No such file or directory

The link has been generated by "npm link" without doing anything special.

Perhaps Serverless should resolve symlinks when it copies files under _tmp?

**Comment by HyperBrain** [comment]
It seems that [fs-extra's copy](https://github.com/jprichardson/node-fs-extra#copy) does not allow any options to explicitly follow symlinks as wrench did.

**Comment by HyperBrain** [comment]
@kennu Could you try with my branch? I set an option. If that works I'll do a PR. For the branch name see the commit mentioned above.

**Comment by kennu** [comment]
@HyperBrain it works! Thanks!

**Comment by eahefnawy** [comment]
@HyperBrain You're awesome! Thanks a lot! :blush:
